File: src/scoring.py
"""Fantasy scoring.

This module is load-bearing: every projection, replacement level and VOR number
downstream is denominated in the points this file produces. It is deliberately a
pure function of a stat line with no I/O, so it can be tested exhaustively.
"""

from __future__ import annotations

import polars as pl

from src.config import FUMBLE_LOST_COLUMNS, SCORING


def score_line(stats: dict, scoring: dict | None = None) -> float:
    """Score a single stat line (a dict of stat -> value).

    Missing stats count as zero, so partial lines (a WR with no passing stats)
    score correctly without the caller having to pad them.
    """
    rules = scoring or SCORING
    total = 0.0
    for stat, points in rules.items():
        total += float(stats.get(stat, 0) or 0) * points
    return total


def add_fantasy_points(
    df: pl.DataFrame,
    scoring: dict | None = None,
    column: str = "fantasy_points_league",
) -> pl.DataFrame:
    """Add a league-scored fantasy points column to an nflverse stats frame.

    Handles the two nflverse quirks that bite here: fumbles lost are split
    across three columns, and any given stats frame may be missing columns
    entirely (older seasons, or position-filtered subsets).
    """
    rules = dict(scoring or SCORING)
    present = set(df.columns)

    # Collapse nflverse's three fumble-lost columns into the single stat the
    # scoring rules name.
    fumble_cols = [c for c in FUMBLE_LOST_COLUMNS if c in present]
    if fumble_cols:
        fumbles = pl.sum_horizontal(
            [pl.col(c).fill_null(0) for c in fumble_cols]
        )
    else:
        fumbles = pl.lit(0.0)

    terms = []
    for stat, points in rules.items():
        if stat == "fumbles_lost":
            terms.append(fumbles * points)
        elif stat in present:
            terms.append(pl.col(stat).fill_null(0) * points)
        # A stat absent from the frame contributes nothing.

    expr = pl.sum_horizontal(terms) if terms else pl.lit(0.0)
    return df.with_columns(expr.cast(pl.Float64).alias(column))

File: v2/lib/scoring.ts
/**
 * Fantasy scoring: a pure dot product of a stat line against a rules vector.
 *
 * Port of `src/scoring.py`, which is already I/O-free and exhaustively tested.
 * Every projection, replacement level and VOR number downstream is denominated
 * in the points this file produces.
 *
 * Missing stats count as zero, so a partial line (a WR with no passing stats)
 * scores correctly without the caller padding it. Note that the Python keeps
 * this tolerance deliberately, and `scripts/export_v2_bundle.py` carries the
 * guard that catches an upstream column rename -- the tolerance is safe only
 * because that guard exists at build time.
 */

export type ScoringRules = Record<string, number>;
export type StatLine = Record<string, number | null | undefined>;

/** Score a single stat line. */
export function scoreLine(stats: StatLine, rules: ScoringRules): number {
  let total = 0;
  for (const stat in rules) {
    const v = stats[stat];
    if (v) total += v * rules[stat];
  }
  return total;
}

/**
 * Score many rows at once, given a columnar table.
 *
 * `columns` maps a stat name to its column index; `rows` are raw arrays as they
 * arrive from the exported bundle.
 */
export function scoreColumnar(
  rows: (number | string | null)[][],
  columns: string[],
  rules: ScoringRules,
): number[] {
  const idx: [number, number][] = [];
  for (const stat in rules) {
    const i = columns.indexOf(stat);
    if (i >= 0) idx.push([i, rules[stat]]);
  }
  return rows.map((row) => {
    let total = 0;
    for (const [i, pts] of idx) {
      const v = row[i];
      if (typeof v === "number") total += v * pts;
    }
    return total;
  });
}

/**
 * Points for a value falling in a banded scale, e.g. points allowed by a
 * defence. Bands are inclusive on both ends; anything unmatched scores zero.
 *
 * Mirrors `_band_expr` in `src/dst.py`.
 */
export function scoreBand(
  value: number,
  bands: { low: number; high: number; points: number }[],
): number {
  for (const b of bands) {
    if (value >= b.low && value <= b.high) return b.points;
  }
  return 0;
}
